### ISSUE TITLE
On MacOS: Illegal Instruction: 4 on CLImage2D test

### ISSUE THREAD

**Issue by arjunmehta** [unbornchikken/NOOOCL#21]
Using `npm 3.10.10` & `node v7.2.1`

When trying to get the tests working on MacOS Sierra, it seems that something has been precompiled with incompatible instruction sets:

Output from `npm test`:
```
  ...
    ✓ should return platforms and devices (42ms)

  CLImage2D
Illegal instruction: 4
```
Info about the `Illegal instruction: 4` error on [StackOverflow](http://stackoverflow.com/questions/14268887/what-is-the-illegal-instruction-4-error-and-why-does-mmacosx-version-min-10).

This happens when testing only CPU as well as GPU.

Doing a bit of debugging, it seems to be happening exactly at [Line 58](https://github.com/unbornchikken/NOOOCL/blob/master/test/imageTests.js#L58). But I'm not sure what to do about it, other than perhaps ensuring that whatever compiled code being used is very platform specific?

**Comment by arjunmehta** [comment]
Bit more debugging this morning.

The script seems to be failing at the `CLEvent.promise` at [Line 58 of imageTests.hs ](https://github.com/unbornchikken/NOOOCL/blob/master/test/imageTests.js#L58)

Further tracing it down to failing exactly at [Line 130 of clEvent.js](https://github.com/unbornchikken/NOOOCL/blob/master/lib/clEvent.js#L130) which seems to be calling `clSetEventCallback`.  Something with it is causing the `Illegal instruction: 4` error and throwing the program.

Let me know if I can try anything to debug.

**Comment by unbornchikken** [comment]
Ok, that's could come from fastcall binary or from the OpenCL binary. Could you try to run unit tests of latest fastcall at your side? If that runs, then we're facing to an OpenCL driver problem. Maybe there are more driver binaries, and we're loading the wrong one or so.

**Comment by arjunmehta** [comment]
I can't build the latest fastcall from scratch as per the issue I posted here: cmake-js/fastcall#11
But I can switch into `NOOOCL/node_modules/fastcall` and run `npm install`.

When I do this I get a few node-gyp warnings:


```
> node-gyp rebuild

gyp WARN download NVM_NODEJS_ORG_MIRROR is deprecated and will be removed in node-gyp v4, please use NODEJS_ORG_MIRROR
gyp WARN download NVM_NODEJS_ORG_MIRROR is deprecated and will be removed in node-gyp v4, please use NODEJS_ORG_MIRROR
gyp WARN download NVM_NODEJS_ORG_MIRROR is deprecated and will be removed in node-gyp v4, please use NODEJS_ORG_MIRROR
  CC(target) Release/obj.target/ffi/deps/libffi/src/prep_cif.o
  CC(target) Release/obj.target/ffi/deps/libffi/src/types.o
  CC(target) Release/obj.target/ffi/deps/libffi/src/raw_api.o
  CC(target) Release/obj.target/ffi/deps/libffi/src/java_raw_api.o
  CC(target) Release/obj.target/ffi/deps/libffi/src/closures.o
  CC(target) Release/obj.target/ffi/deps/libffi/src/x86/ffi.o
  CC(target) Release/obj.target/ffi/deps/libffi/src/x86/ffi64.o
  CC(target) Release/obj.target/ffi/deps/libffi/src/x86/darwin.o
  CC(target) Release/obj.target/ffi/deps/libffi/src/x86/darwin64.o
  LIBTOOL-STATIC Release/libffi.a
  CXX(target) Release/obj.target/ffi_bindings/src/ffi.o
../src/ffi.cc:111:11: warning: 'ForceSet' is deprecated [-Wdeprecated-declarations]
  target->ForceSet(Nan::New<String>("RTLD_NEXT").ToLocalChecked(), WrapPointer((char *)RTLD_NEXT), static_cast<PropertyAttribute>(ReadOnly | DontDelete));
          ^
/me/.node-gyp/6.9.2/include/node/v8.h:2695:22: note: 'ForceSet' has been explicitly marked deprecated here
                bool ForceSet(Local<Value> key, Local<Value> value,
                     ^
../src/ffi.cc:114:11: warning: 'ForceSet' is deprecated [-Wdeprecated-declarations]
  target->ForceSet(Nan::New<String>("RTLD_DEFAULT").ToLocalChecked(), WrapPointer((char *)RTLD_DEFAULT), static_cast<PropertyAttribute>(ReadOnly | DontDelete));
          ^
/me/.node-gyp/6.9.2/include/node/v8.h:2695:22: note: 'ForceSet' has been explicitly marked deprecated here
                bool ForceSet(Local<Value> key, Local<Value> value,
                     ^
../src/ffi.cc:117:11: warning: 'ForceSet' is deprecated [-Wdeprecated-declarations]
  target->ForceSet(Nan::New<String>("RTLD_SELF").ToLocalChecked(), WrapPointer((char *)RTLD_SELF), static_cast<PropertyAttribute>(ReadOnly|DontDelete));
          ^
/me/.node-gyp/6.9.2/include/node/v8.h:2695:22: note: 'ForceSet' has been explicitly marked deprecated here
                bool ForceSet(Local<Value> key, Local<Value> value,
                     ^
../src/ffi.cc:120:11: warning: 'ForceSet' is deprecated [-Wdeprecated-declarations]
  target->ForceSet(Nan::New<String>("RTLD_MAIN_ONLY").ToLocalChecked(), WrapPointer((char *)RTLD_MAIN_ONLY), static_cast<PropertyAttribute>(ReadOnly|DontDelete));
          ^
/me/.node-gyp/6.9.2/include/node/v8.h:2695:22: note: 'ForceSet' has been explicitly marked deprecated here
                bool ForceSet(Local<Value> key, Local<Value> value,
                     ^
../src/ffi.cc:123:11: warning: 'ForceSet' is deprecated [-Wdeprecated-declarations]
  target->ForceSet(Nan::New<String>("FFI_ARG_SIZE").ToLocalChecked(), Nan::New<Uint32>((uint32_t)sizeof(ffi_arg)), static_cast<PropertyAttribute>(ReadOnly|DontDelete));
          ^
/me/.node-gyp/6.9.2/include/node/v8.h:2695:22: note: 'ForceSet' has been explicitly marked deprecated here
                bool ForceSet(Local<Value> key, Local<Value> value,
                     ^
../src/ffi.cc:124:11: warning: 'ForceSet' is deprecated [-Wdeprecated-declarations]
  target->ForceSet(Nan::New<String>("FFI_SARG_SIZE").ToLocalChecked(), Nan::New<Uint32>((uint32_t)sizeof(ffi_sarg)), static_cast<PropertyAttribute>(ReadOnly | DontDelete));
          ^
/me/.node-gyp/6.9.2/include/node/v8.h:2695:22: note: 'ForceSet' has been explicitly marked deprecated here
                bool ForceSet(Local<Value> key, Local<Value> value,
                     ^
../src/ffi.cc:125:11: warning: 'ForceSet' is deprecated [-Wdeprecated-declarations]
  target->ForceSet(Nan::New<String>("FFI_TYPE_SIZE").ToLocalChecked(), Nan::New<Uint32>((uint32_t)sizeof(ffi_type)), static_cast<PropertyAttribute>(ReadOnly | DontDelete));
          ^
/me/.node-gyp/6.9.2/include/node/v8.h:2695:22: note: 'ForceSet' has been explicitly marked deprecated here
                bool ForceSet(Local<Value> key, Local<Value> value,
                     ^
../src/ffi.cc:126:11: warning: 'ForceSet' is deprecated [-Wdeprecated-declarations]
  target->ForceSet(Nan::New<String>("FFI_CIF_SIZE").ToLocalChecked(), Nan::New<Uint32>((uint32_t)sizeof(ffi_cif)), static_cast<PropertyAttribute>(ReadOnly | DontDelete));
          ^
/me/.node-gyp/6.9.2/include/node/v8.h:2695:22: note: 'ForceSet' has been explicitly marked deprecated here
                bool ForceSet(Local<Value> key, Local<Value> value,
                     ^
../src/ffi.cc:132:11: warning: 'ForceSet' is deprecated [-Wdeprecated-declarations]
  target->ForceSet(Nan::New<String>("HAS_OBJC").ToLocalChecked(), Nan::New<Boolean>(hasObjc), static_cast<PropertyAttribute>(ReadOnly | DontDelete));
          ^
/me/.node-gyp/6.9.2/include/node/v8.h:2695:22: note: 'ForceSet' has been explicitly marked deprecated here
                bool ForceSet(Local<Value> key, Local<Value> value,
                     ^
9 warnings generated.
  CXX(target) Release/obj.target/ffi_bindings/src/callback_info.o
  CXX(target) Release/obj.target/ffi_bindings/src/threaded_callback_invokation.o
  SOLINK_MODULE(target) Release/ffi_bindings.node
clang: warning: libstdc++ is deprecated; move to libc++ with a minimum deployment target of OS X 10.9
ld: warning: could not create compact unwind for _ffi_call_unix64: does not use RBP or RSP based frame
```

But it seems to build okay. I run `npm test` and all 97 tests pass.

**Comment by unbornchikken** [comment]
fastcall has nothing to do with node-gyp actually, it's based on CMake.js. Only the developer mode build uses node-gyp to build [ffi](https://github.com/node-ffi/node-ffi) and [ref](https://github.com/TooTallNate/ref) just for the benchmarks. I've added instructions [there](https://github.com/cmake-js/fastcall/blob/master/README.md#clone-and-build).

But, since you did run fastcall tests, and there wasn't any problem on your rig, we have a driver issue there. Which is very unfortunate, because I really cannot do anything to help. You get illegal instruction because OpenCL driver is compiled with SSE4 or AVX instruction set, and your hardware doesn't support it. Maybe there is a driver update available or something, I don't know Macs. Maybe there is another OpenCL binary in your system directory that should be loaded instead of that one from your latest PR?

**Comment by unbornchikken** [comment]
I can reproduce this on a Mac with AMD GPU. However in my case even the CLBuffer get stuck at the first phase. If I switch from GPU to CPU CLBuffer tests pass, but I get "Illegal instuction: 4". Worth reading though: http://www.blendernation.com/2015/05/04/an-open-letter-to-tim-cook/

Looks like Tim are still giving a pile of shit on this. @wuxianliang, could you try tests of NOOOCL@latest on your Mac?

**Comment by unbornchikken** [comment]
It looks like, the default device is an Intel one on the Mac that I have access to, so it is an Intel driver issue definitely. @arjunmehta Do you have clinfo? Is it same like https://github.com/unbornchikken/NOOOCL/issues/18?

**Comment by arjunmehta** [comment]
Yes same [output from `clinfo` on my Late 2013 13" Macbook Pro](https://github.com/unbornchikken/NOOOCL/files/653837/clinfo.txt)

**Comment by unbornchikken** [comment]
I've dug myself into articles about this issue, and I believe this is really originated from fastcall. It uses "-march=native" which could cause issues on older Macs. I think if I add a unit test to fastcall which uses the same argument scheme like OpenCL event callback, it will crash with Illegal instruction. Or there might be issues about the callback which might get garbage collected on Mac before the callback really happens, this could lead to an Illegal instruction. Or there might be issues with the assembler arguments, those are kinda special case on Mac.

So I have some suspects, but I don't have access to a Mac until next week.

**Comment by arjunmehta** [comment]
If you want, You can use Travis.

Set up a branch and configure for OS X. My PR for NOOOCL should give you an idea to get started with your config but let me know if you have any questions!

**Comment by unbornchikken** [comment]
Ok, I managed to debug that, and unfortunately it's from the OpenCL driver. The crash originated from the driver's `uint8_rgba.dylib`. It throws by executing `ud2` instruction in a macro called: `__illegal_sampler_mode__integer_coords_must_use_nearest_filter_mode`. Dunno why, but thats the problem. Gotta look into it.

**Comment by unbornchikken** [comment]
Intel driver doesn't understand linear interpolation when using integer coordinates. Dunno why, it should work by that way, and of course works on all other platfroms.

The issue fixed as of v0.12.4, all test should pass with Intel drivers.

**Comment by arjunmehta** [comment]
Nice! It seems to be working/passing the tests. Great job!
